core/src/main/java/org/ehcache/core/statistics/TierOperationStatistic.java

```diff
@@ -0,0 +1,231 @@
+  }
+
+  @Override
+  public void addDerivedStatistic(final ChainedOperationObserver<? super D> derived) {
+    operationStatistic.addDerivedStatistic(new ChainedOperationObserver<S>() {
+      @Override
+      public void begin(long time) {
+        derived.begin(time);
+      }
+
+      @Override
+      public void end(long time, S result) {
+        derived.end(time, (D) result);
+      }
+
+      @Override
+      public void end(long time, S result, long... parameters) {
+        derived.end(time, (D) result, parameters);
+      }
+    });
+  }
+
+  @Override
+  public void removeDerivedStatistic(ChainedOperationObserver<? super D> derived) {
+    operationStatistic.removeDerivedStatistic((ChainedOperationObserver<? super S>) derived);
+  }
+
+  @Override
+  public void begin() {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void end(D result) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void end(D result, long... parameters) {
+    throw new UnsupportedOperationException();
+  }
+
+  private static String findDiscriminator(Object rootNode) {
+    Set<TreeNode> results = queryBuilder().chain(self())
+        .children().filter(
+            context(attributes(Matchers.<Map<String, Object>>allOf(
+                hasAttribute("discriminator", new Matcher<Object>() {
+                  @Override
+                  protected boolean matchesSafely(Object object) {
+                    return object instanceof String;
+                  }
+                }))))).build().execute(Collections.singleton(ContextManager.nodeFor(rootNode)));
+
+    if (results.size() > 1) {
+      throw new IllegalStateException("More than one discriminator attribute found");
+    } else if (results.isEmpty()) {
+      return null;
+    } else {
+      TreeNode node = results.iterator().next();
+      return (String) node.getContext().attributes().get("discriminator");
+    }
+  }
+
+  private static OperationStatistic findOperationStat(Object rootNode, final String statName) {
+    Query q = queryBuilder().chain(self())
+        .descendants().filter(context(identifier(subclassOf(OperationStatistic.class)))).build();
+
+    Set<TreeNode> operationStatisticNodes = q.execute(Collections.singleton(ContextManager.nodeFor(rootNode)));
+    Set<TreeNode> result = queryBuilder()
+        .filter(
+            context(attributes(Matchers.<Map<String, Object>>allOf(
+                hasAttribute("name", statName))))).build().execute(operationStatisticNodes);
+
+    if (result.size() != 1) {
+      throw new RuntimeException("a single stat was expected; found " + result.size());
+    }
+
+    TreeNode node = result.iterator().next();
+    return (OperationStatistic) node.getContext().attributes().get("this");
+  }
+
+  public static <X> Set<X> set(X... xs) {
+    return new HashSet<X>(Arrays.asList(xs));
+  }
+
+  public static class TierOperationOutcomes {
+
+    public enum GetOutcome {
```

Comment:
i need to double check this one.  I didn't think it was actually needed but let me make sure.